pisa/stages/flux/pi_mceq_barr_red.py

```diff
@@ -0,0 +1,489 @@
+        delta_index = self.params.delta_index.value.m_as("dimensionless")
+        energy_pivot = self.params.energy_pivot.value.m_as("GeV")
+
+        # Grab the pion ratio
+        pion_ratio = self.params.pion_ratio.value.m_as("dimensionless")
+
+        # Map the user parameters into the Barr +/- params
+        # pi- production rates is restricted by the pi-ratio, just as in arXiv:0611266
+        # TODO might want dedicated priors for pi- params (but without corresponding free params)
+        gradient_params_mapping = collections.OrderedDict()
+        gradient_params_mapping["a_f+"] = self.params.barr_a_f_Pi.value.m_as("dimensionless")
+        gradient_params_mapping["g+"] = self.params.barr_g_Pi.value.m_as("dimensionless")
+        gradient_params_mapping["h+"] = self.params.barr_h_Pi.value.m_as("dimensionless")
+        gradient_params_mapping["i+"] = self.params.barr_i_Pi.value.m_as("dimensionless")
+        for k in list(gradient_params_mapping.keys()):
+            gradient_params_mapping[k.replace("+", "-")] = self.antipion_production(
+                gradient_params_mapping[k], pion_ratio
+            )
+
+        # kaons
+        # as the kaon ratio is unknown, K- production is not restricted
+        gradient_params_mapping["w+"] = self.params.barr_w_K.value.m_as("dimensionless")
+        gradient_params_mapping["w-"] = self.params.barr_w_antiK.value.m_as("dimensionless")
+        gradient_params_mapping["x+"] = self.params.barr_x_K.value.m_as("dimensionless")
+        gradient_params_mapping["x-"] = self.params.barr_x_antiK.value.m_as("dimensionless")
+        gradient_params_mapping["y+"] = self.params.barr_y_K.value.m_as("dimensionless")
+        gradient_params_mapping["y-"] = self.params.barr_y_antiK.value.m_as("dimensionless")
+        gradient_params_mapping["z+"] = self.params.barr_z_K.value.m_as("dimensionless")
+        gradient_params_mapping["z-"] = self.params.barr_z_antiK.value.m_as("dimensionless")
+
+        # Populate array Barr param array
+        for (
+            gradient_param_name,
+            gradient_param_idx,
+        ) in self.gradient_param_indices.items():
+            assert gradient_param_name in gradient_params_mapping, (
+                "Gradient parameter '%s' missing from mapping" % gradient_param_name
+            )
+            self.gradient_params[gradient_param_idx] = gradient_params_mapping[
+                gradient_param_name
+            ]
+
+        #
+        # Loop over containers
+        #
+
+        for container in self.data:
+
+            #
+            # Apply the systematics to the flux
+            #
+
+            nubar = container["nubar"]
+            if nubar > 0: flux_key = "nu_flux_nominal"
+            elif nubar < 0: flux_key = "nubar_flux_nominal"
+
+            apply_sys_vectorized(
+                container["true_energy"].get(WHERE),
+                container["true_coszen"].get(WHERE),
+                delta_index,
```

Comment:
can you put an `FTYPE()` around these two scalars for explicit typecasting? (just because i'm paranoid)